mmengine/runner/loops.py

```diff
@@ -281,6 +281,7 @@ def run(self) -> None:
             if (self.runner.val_loop is not None
                     and self._iter >= self.val_begin
                     and self._iter % self.val_interval == 0):
+                self.runner.call_hook('after_train_epoch')
```

Comment:
This line will call all the hooks, which is not expected.
We can implement a `before_val_epoch` method in [SyncBuffersHook](https://github.com/open-mmlab/mmengine/blob/main/mmengine/hooks/sync_buffer_hook.py) which does the same action with `after_train_epoch`.
In addition, we can also add an attribute `called_in_train` to `SyncBuffersHook` which will be set as True in `after_train_epoch` after `synchronization` and set as False in `before_val_epoch`.

Reply:
@zhouzaida Thank you for such detailed explanation. May I ask the purpose of adding the attribute `called_in_train`, as it has not been used anywhere.

Reply:
Adding `called_in_train` is to avoid synchronizing the buffers twice in EpochBasedTrainLoop which will sync buffers in `after_train_epoch`.

https://github.com/open-mmlab/mmengine/blob/aeb5c454c56212eb1f94c5732de9be747d85ac87/mmengine/runner/loops.py#L107-L114

```python
@HOOKS.register_module()
class SyncBuffersHook(Hook):
    """Synchronize model buffers such as running_mean and running_var in BN at
    the end of each epoch."""

    priority = 'NORMAL'

    def __init__(self) -> None:
        self.distributed = is_distributed()
        self.called_in_train = False

    def after_train_epoch(self, runner) -> None:
        """All-reduce model buffers at the end of each epoch.
        Args:
            runner (Runner): The runner of the training process.
        """
        if self.distributed:
            all_reduce_params(runner.model.buffers(), op='mean')
            self.called_in_train = True
            

    def before_val_epoch(self, runner) -> None:
        """All-reduce model buffers at the end of each epoch.
        Args:
            runner (Runner): The runner of the training process.
        """
        if self.distributed and not self.called_in_train:
            all_reduce_params(runner.model.buffers(), op='mean')
            self.called_in_train = False
```

Reply:
@zhouzaida Thanks! I have made changes accordingly. 
There is only one thing left to discuss. For the implementation of `before_val_epoch`, I made `self.called_in_train = False` executed whatever the `called_in_train` is. Please correct me if I am mistaken.